typescript/solver/solvers/BaseListener.ts

```diff
@@ -0,0 +1,68 @@
+import type { Provider } from "@ethersproject/providers";
+import { chainMetadata } from "@hyperlane-xyz/registry";
+import { MultiProvider } from "@hyperlane-xyz/sdk";
+import type { Contract, Signer } from "ethers";
+
+import type { Logger } from "../logger.js";
+import type { TypedEvent, TypedListener } from "../typechain/common.js";
+
+export abstract class BaseListener<
+  TContract extends Contract,
+  TEvent extends TypedEvent,
+  TParsedArgs,
+> {
+  protected constructor(
+    private readonly contractFactory: {
+      connect(address: string, signerOrProvider: Signer | Provider): TContract;
+    },
+    private readonly eventName: Extract<keyof TContract["filters"], string>,
+    private readonly metadata: {
+      address: string;
+      chainId: number;
+      solverName: string;
```

Comment:
IMO, we shouldn't call this `solverName`, but more like `intentProtocolType` or something like that? Also let's make an enum out of this type?